### ISSUE TITLE
[pt2] Python float argument that changes between calls goes stale in the cached graph (residual of #194976): norm `eps`, `huber_loss` delta, SDPA `scale`, `interpolate(scale_factor=...)` keep an earlier value (`interpolate`: old output shape)

### ISSUE THREAD

**Issue by cuiliaomei-beep** [pytorch/pytorch#197889]
### 🐛 Describe the bug

**Summary.** For `f(x, s)` that uses a Python float as `x * s` and, through arithmetic, as a float parameter of an operator (`F.rms_norm(x * s, (4,), eps=s + 1e-3)`), eager and `aot_eager` use the current `s`, while the default `torch.compile` keeps the parameter computed from the second `s` for all later values (torch 2.14.0 and nightly 2026-09-21, CPU and CUDA). A schedule that varies a loss `delta`, a `label_smoothing`, an attention `scale` or an `interpolate` scale factor computes with an old value, with no error, and `interpolate` returns the old shape. Because fake-tensor propagation specialises the symbolic float with a ShapeEnv guard rather than a graph input, AOT bakes the constant while the Dynamo graph stays identical for every `s`, so after each recompile AOTAutogradCache returns the first artifact.

```python
import torch, torch.nn.functional as F

torch.manual_seed(0)
x, y = torch.randn(4, 4), torch.randn(4, 4)
programs = {
    "rms_norm(x * s, eps=s + 1e-3)":          lambda s: F.rms_norm(x * s, (4,), eps=s + 1e-3),
    "huber_loss(x * s, y, delta=s + 0.1)":    lambda s: F.huber_loss(x * s, y, delta=s + 0.1),
    "interpolate(x * s, scale_factor=1 + s)": lambda s: F.interpolate((x * s)[None], scale_factor=1.0 + s),
}
for name, f in programs.items():
    torch._dynamo.reset()
    cf = torch.compile(f)
    print(name)
    for s in (0.9, 0.5, 0.25, 2.0):
        want, got = f(s), cf(s)
        same = want.shape == got.shape and torch.allclose(want, got)
        print(f"    s={s:<5} eager shape {tuple(want.shape)}  compiled shape {tuple(got.shape)}  " + ("ok" if same else "WRONG" + ("" if want.shape != got.shape else f" (max abs diff {(want - got).abs().max().item():.3f})")))
```

```
rms_norm(x * s, eps=s + 1e-3)
    s=0.9   eager shape (4, 4)  compiled shape (4, 4)  ok
    s=0.5   eager shape (4, 4)  compiled shape (4, 4)  ok
    s=0.25  eager shape (4, 4)  compiled shape (4, 4)  WRONG (max abs diff 0.218)
    s=2.0   eager shape (4, 4)  compiled shape (4, 4)  WRONG (max abs diff 0.348)
huber_loss(x * s, y, delta=s + 0.1)
    s=0.9   eager shape ()  compiled shape ()  ok
    s=0.5   eager shape ()  compiled shape ()  ok
    s=0.25  eager shape ()  compiled shape ()  WRONG (max abs diff 0.120)
    s=2.0   eager shape ()  compiled shape ()  WRONG (max abs diff 0.731)
interpolate(x * s, scale_factor=1 + s)
    s=0.9   eager shape (1, 4, 7)  compiled shape (1, 4, 7)  ok
    s=0.5   eager shape (1, 4, 6)  compiled shape (1, 4, 6)  ok
    s=0.25  eager shape (1, 4, 5)  compiled shape (1, 4, 6)  WRONG
    s=2.0   eager shape (1, 4, 12)  compiled shape (1, 4, 6)  WRONG
```

The first call specialises `s` (static), the second makes it automatically dynamic and is still right, and every value after that reuses the parameter of the second call: for `rms_norm` the compiled result equals eager's result with `eps = 0.5 + 1e-3` exactly. With `TORCHINDUCTOR_FX_GRAPH_CACHE=0` or `TORCHINDUCTOR_AUTOGRAD_CACHE=0` (either one), and with `backend="aot_eager"`, all rows are `ok`.

Of 51 operators with a float parameter that I tried in this shape, 18 are affected on nightly: `rms_norm`, `layer_norm`, `group_norm`, `batch_norm`, `instance_norm` (eps) and `batch_norm` momentum, `cosine_similarity` eps, `pairwise_distance` / `cdist` p, `quantile` q, `interpolate` scale_factor (nearest and linear), `smooth_l1_loss` beta, `huber_loss` delta, `cross_entropy` label_smoothing, `poisson_nll_loss` eps, `triplet_margin_loss` margin, and `scaled_dot_product_attention` scale. Pointwise ops whose scalar is tensorified (`clamp`, `leaky_relu`, `elu`, `hardtanh`, `threshold`, `addcmul`, `lerp`, `where`, `pow`, …) are correct on nightly; on 2.14.0, which predates the fix for #194976, many of those are wrong as well (`torch.add(x * s, y, alpha=1 - s)`, the EMA update, among them).

**Root cause.** `TORCH_LOGS=graph_code,aot_graphs,recompiles` for the `rms_norm` program:

```
[0/1] Dynamo graph:   item = l_s_.item(); mul = l_x_ * item; add = item + 0.001; rms_norm = torch.rms_norm(mul, (4,), None, add)
[0/1] AOT graph:      mul = aten.mul.Tensor(arg0_1, _local_scalar_dense) ... add_1 = aten.add.Scalar(mean, 0.5009999871253967)
[0/2] recompile:      guard failure  (0.001 + s) == 0.501   # functional.py: rms_norm (_dynamo/utils.py in run_node)
[0/2] Dynamo graph:   identical text to [0/1]
[0/2] AOT graph:      "Forward graph (from cache)" ... add_1 = aten.add.Scalar(mean, 0.5009999871253967)
```

The operator takes a C++ `double`, so tracing it on fake tensors specialises the expression `zf + 0.001` and records the ShapeEnv guard `(0.001 + s) == 0.501`; the value is burned into the AOT graph, but the Dynamo graph still carries the symbolic expression. The guard works (the third call recompiles), but the recompile produces the same Dynamo graph, AOTAutogradCache keys on that graph and prunes guards over symbols that are not placeholder SymInts, and the artifact with `0.501` inside comes back. This is the mechanism described in #195040, which closed it for the final sweep of `_tensorify_python_scalars.py`; specialisation that happens earlier, during Dynamo's fake-tensor propagation of an op with a float schema argument, is not covered.

**Possible fix.** Treat a float symbol that gets specialised during tracing the same way #195040 treats one specialised in the tensorify sweep: mark it with `TensorifyState.specialize` and restart analysis so that Dynamo bakes the value into its own output graph (the value then is part of the cache key); or, more narrowly, make AOTAutogradCache bypass / miss when the ShapeEnv holds an equality guard over a float symbol that is not a graph input.

### Error logs

No error. The recompile log shows the guard failing and the forward graph being served from cache.

### Versions

```
PyTorch version: 2.14.0+cpu
OS: Microsoft Windows 11 Pro for Workstations (10.0.26200 64-bit)
MSVC: Microsoft (R) C/C++ Optimizing Compiler Version 19.44.35228 for x64
Python version: 3.14.7 (64-bit runtime)
Is CUDA available: False
CPU: 11th Gen Intel(R) Core(TM) i9-11900T @ 1.50GHz
[pip3] numpy==2.5.2
[pip3] torch==2.14.0+cpu
```

Same result on nightly (2.15.0.dev20260911 on Windows, 2.15.0.dev20260921 on Linux), on Linux CPU with 2.14.0, and on CUDA (Tesla T4, 2.14.0+cu130 and nightly 2.15.0.dev20260921+cu130).

cc @chauhang @penguinwu @ezyang @bobrenjc93 @aditvenk @laithsakka @bdhirsh @aorenste

**Comment by cuiliaomei-beep** [comment]
A second entry point with the same root cause: **integer** arguments computed from the float. `int(s * k)` on the dynamic float is a symbolic integer; operators that accept a SymInt there are fine (`torch.zeros(int(s*8))`, slicing, `F.pad`, `repeat`, `roll`, `topk`, `narrow`, `tril` all follow `s`), but operators whose schema takes a plain `int` specialise it during tracing, and from the third distinct `s` on the cached artifact keeps the value of the second call:

```python
import torch, torch.nn.functional as F

torch.manual_seed(0)
x = torch.randn(8, 8)
programs = {
    "cumsum(dim=int(s*2) % 2)":            lambda s: (x * s).cumsum(int(s * 2) % 2),
    "round(decimals=int(s*4))":            lambda s: (x * s).round(decimals=int(s * 4)),
    "linspace(0, 1, int(s*8) + 2)":        lambda s: x[0, 0] * s + torch.linspace(0, 1, int(s * 8) + 2),
    "one_hot(num_classes=int(s*8) + 3)":   lambda s: F.one_hot(torch.tensor([0, 1, 2]), int(s * 8) + 3) * (x[0, 0] * s),
    "adaptive_avg_pool1d(int(s*6) + 1)":   lambda s: F.adaptive_avg_pool1d((x * s)[None], int(s * 6) + 1),
    "unfold(0, int(s*4) + 1, 1)":          lambda s: (x * s).unfold(0, int(s * 4) + 1, 1),
}
for name, f in programs.items():
    torch._dynamo.reset()
    cf = torch.compile(f)
    row = []
    for s in (0.9, 0.5, 0.25, 1.0):
        want, got = f(s), cf(s)
        row.append("ok" if want.shape == got.shape and torch.allclose(want, got) else
                   (f"shape {tuple(got.shape)} != {tuple(want.shape)}" if want.shape != got.shape else f"max abs diff {(want - got).abs().max().item():.3g}"))
    print(f"{name:36s} " + " | ".join(f"s={s}: {r}" for s, r in zip((0.9, 0.5, 0.25, 1.0), row)))
```

```
cumsum(dim=int(s*2) % 2)             s=0.9: ok | s=0.5: ok | s=0.25: max abs diff 1.79 | s=1.0: max abs diff 7.18
round(decimals=int(s*4))             s=0.9: ok | s=0.5: ok | s=0.25: max abs diff 0.05 | s=1.0: max abs diff 0.0049
linspace(0, 1, int(s*8) + 2)         s=0.9: ok | s=0.5: ok | s=0.25: shape (6,) != (4,) | s=1.0: shape (6,) != (10,)
one_hot(num_classes=int(s*8) + 3)    s=0.9: ok | s=0.5: ok | s=0.25: shape (3, 7) != (3, 5) | s=1.0: shape (3, 7) != (3, 11)
adaptive_avg_pool1d(int(s*6) + 1)    s=0.9: ok | s=0.5: ok | s=0.25: shape (1, 8, 4) != (1, 8, 2) | s=1.0: shape (1, 8, 4) != (1, 8, 7)
unfold(0, int(s*4) + 1, 1)           s=0.9: ok | s=0.5: ok | s=0.25: shape (6, 8, 3) != (7, 8, 2) | s=1.0: shape (6, 8, 3) != (4, 8, 5)
```

`cumsum` accumulates along the wrong dimension and `round` uses the wrong number of decimals with no shape change to give it away. As in the description, either `TORCHINDUCTOR_FX_GRAPH_CACHE=0` or `TORCHINDUCTOR_AUTOGRAD_CACHE=0` makes every row `ok`. So a fix at the cache level (or a restart-and-specialise for any symbol whose *derived* expression gets specialised, integer-valued ones included) would cover both families; a fix limited to float schema arguments would leave these.

For completeness, value-dependent decompositions are not affected: `pow` with exponent 2 / 0.5 / 0, `addmm` / `baddbmm` with `beta=0` on a NaN input, `addcmul(value=0)`, `lerp`, `dropout` with p in {0, 1}, `clamp` with min > max, `softplus`, `norm` / `normalize` p all track `s` correctly.
